src/env/double_write.rs

```diff
@@ -0,0 +1,722 @@
+                let fd = Arc::new(Self::resolve(file2));
+                fd2.write().unwrap().replace(fd.clone());
+                for (task, cb) in rx2 {
+                    if task == FileTask::Stop {
+                        break;
+                    }
+                    let res = Self::process(&fd, task);
+                    counter2.fetch_add(1, Ordering::Relaxed);
+                    cb(res);
+                }
+            })
+        };
+        Self {
+            disk1: tx1,
+            disk2: tx2,
+            counter1,
+            counter2,
+            fd1,
+            fd2,
+            handle1: Some(handle1),
+            handle2: Some(handle2),
+        }
+    }
+
+    fn resolve(
+        file: Either<oneshot::Receiver<IoResult<Option<F::Handle>>>, F::Handle>,
+    ) -> F::Handle {
+        match file {
+            Either::Left(f) => {
+                // TODO: should we handle the second disk io error
+                block_on(f).unwrap().unwrap().unwrap()
+            }
+            Either::Right(fd) => fd,
+        }
+    }
+
+    fn process(fd: &Arc<F::Handle>, task: FileTask) -> IoResult<TaskRes<F>> {
+        match task {
+            FileTask::Truncate(offset) => fd.truncate(offset).map(TaskRes::Truncate),
+            FileTask::FileSize => fd.file_size().map(|s| TaskRes::FileSize(s)),
+            FileTask::Sync => fd.sync().map(TaskRes::Sync),
+            FileTask::Write { offset, bytes } => fd.write(offset, &bytes).map(|s| TaskRes::Write(s)),
+            FileTask::Allocate { offset, size } => fd.allocate(offset, size).map(TaskRes::Allocate),
+            FileTask::NewWriter { fs } => fs.new_writer(fd.clone()).map(|w| TaskRes::NewWriter(w)),
+            FileTask::Stop => unreachable!(),
+        }
+    }
+
+    fn new_writer(&self) -> IoResult<HedgedWriter<F>> {
+        let task1 = FileTask::NewWriter(self.base.clone());
+        let task2 = FileTask::NewWriter(self.base.clone());
+        let (cb1, mut f1) = paired_future_callback();
+        let (cb2, mut f2) = paired_future_callback();
+        self.disk1.send((task1, cb1)).unwrap();
+        self.disk2.send((task2, cb2)).unwrap();
+
+        select! {
+            res1 = f1 => res1.unwrap().map(|w| HedgedWriter::new(self.disk1.clone(), self.disk2.clone(), 
+                Either::Right(w.unwrap()), Either::Left(f2) )),
+            res2 = f2 => res2.unwrap().map(|w| HedgedHandle::new(
+                Either::Left(f1), Either::Right(w.unwrap()) )),
+        }
+        select! {
+            res1 = f1 => res1.unwrap().map(|_| ()),
+            res2 = f2 => res2.unwrap().map(|_| ()),
+        }
+    }
+
+    fn read(&self, offset: usize, buf: &mut [u8]) -> IoResult<usize> {
+        // TODO: read simultaneously from both disks
+        // choose latest to perform read
+        let count1 = self.counter1.load(Ordering::Relaxed);
+        let count2 = self.counter2.load(Ordering::Relaxed);
+        match count1.cmp(&count2) {
```

Comment:
There are 3 steps here before reading a file: 
1. get counter1
2. get counter2
3. compare counter1 and counter2

The 3 steps together are not atomic. If we get an wrong result that counter1 < counter2 due to the in-atomicity, and the offset > file size for counter1, what will happen?

Reply:
It doesn't matter.
Let's say disk2 is the slower one, and its file size < offset
- first get s1 = counter1, we can know that s1 >= offset >= file size,
- then get s2 = counter2, 
  - if s1 <= s2, we read on disk2, because s2 >= s1 >= offset >= file size, disk2 must have the data now
  - if s1 > s2, we read on disk1

Reply:
My overall impression of this PR is that we are leveraging certain undocumented invariants, e.g.  sole append-only writer plus multi-readers,  in the upper layers to ensure the safety of concurrency in the lower-level HedgedFileSystem. Two drawbacks I can think of
* There are many variants in the FileSystem layer,  for example the read-write conflicts,  in-atomicity of operations, etc. It's complicated to cover all the cases, and it make the code hard to be understood.
* In the future, if any engineer is unaware of these hidden relationships, it could easily lead to significant disasters.